internal/app/app.go

```diff
@@ -104,17 +107,55 @@ func NewListener(c *cli.Context) (*listener.Listener, error) {
 	blockExpiration := c.Duration(blockExpirationFlag.Name)
 	blockKeeper := block.NewRedisBlockKeeper(l, redisClient, maxNumBlocks, blockExpiration)
 
-	maxLen := c.Int64(publisherMaxLenFlag.Name)
-	redisStream := redis.NewStream(redisClient, maxLen)
-
 	topic := c.String(publisherTopicFlag.Name)
-	handler := listener.NewHandler(l, topic, httpEVMClient, blockKeeper, redisStream,
+	publisher, err := getPublisher(c, redisClient, topic)
+	if err != nil {
+		l.Errorw("Fail to get publisher", "error", err)
+
+		return nil, err
+	}
+	encoder := getMessageEncoder(c)
+
+	handler := listener.NewHandler(l, topic, httpEVMClient, blockKeeper, publisher, encoder,
 		listener.WithEventLogs(nil, nil))
 
 	return listener.New(l, wsEVMClient, httpEVMClient, handler, sanityEVMClient, sanityCheckInterval,
 		listener.WithEventLogs(nil, nil)), nil
 }
 
+func getPublisher(c *cli.Context, redisClient *redis.Client, topic string) (publisherpkg.Publisher, error) {
+	var publisher publisherpkg.Publisher
+	var err error
+
+	publisherType := c.String(publisherTypeFlag.Name)
+	switch publisherType {
+	case publisherpkg.PublisherTypeKafka:
+		addresses := c.StringSlice(kafkaAddrsFlag.Name)
+		publisher, err = kafka.NewPublisher(&kafka.Config{Addresses: addresses})
+		if err != nil {
+			return nil, err
+		}
+		if err := kafka.ValidateTopicName(topic); err != nil {
+			return nil, err
+		}
+	default:
+		maxLen := c.Int64(publisherMaxLenFlag.Name)
+		publisher = redis.NewStream(redisClient, maxLen)
+	}
+
+	return publisher, err
+}
+
+func getMessageEncoder(c *cli.Context) encoder.Encoder {
+	publisherType := c.String(publisherTypeFlag.Name)
+	switch publisherType {
```

Comment:
I think message encoder should not depend on type of publisher.

Reply:
I added an `encoderType` flag.